ee/clickhouse/models/property.py

```diff
@@ -225,50 +227,60 @@ def get_property_values_for_key(key: str, team: Team, value: Optional[str] = Non
-        if len(or_conditions) > 0:
-            conditions.append("(" + (" OR ".join(or_conditions)) + ")")
+            conditions.append(f"match(elements_chain, %({key})s)")
+        if len(conditions) > 0:
+            conditions.append("(" + (" OR ".join(conditions)) + ")")
 
     if filters.get("tag_name"):
-        or_conditions = []
         tag_names = filters["tag_name"] if isinstance(filters["tag_name"], list) else [filters["tag_name"]]
         for idx, tag_name in enumerate(tag_names):
             key = f"{prepend}_{idx}_tag_name_regex"
             params[key] = rf"(^|;){tag_name}(\.|$|;|:)"
-            or_conditions.append(f"match(elements_chain, %({key})s)")
-        if len(or_conditions) > 0:
-            conditions.append("(" + (" OR ".join(or_conditions)) + ")")
+            conditions.append(f"match(elements_chain, %({key})s)")
+        if len(conditions) > 0:
+            conditions.append("(" + (" OR ".join(conditions)) + ")")
 
     attributes: Dict[str, List] = {}
-
     for key in ["href", "text"]:
-        vals = filters.get(key)
-        if filters.get(key):
-            attributes[key] = [re.escape(vals)] if isinstance(vals, str) else [re.escape(text) for text in filters[key]]
-
-    if len(attributes.keys()) > 0:
-        or_conditions = []
+        if vals := filters.get(key):
+            if operator.endswith("_set"):
+                attributes[key] = [r".*"]
+            else:
+                vals = [vals] if isinstance(vals, str) else vals
+                if operator.endswith("icontains"):
+                    attributes[key] = [f".*{re.escape(text)}.*" for text in vals]
+                elif operator.endswith("regex"):
+                    attributes[key] = vals
+                else:
+                    attributes[key] = [re.escape(text) for text in vals]
+    is_negated = operator in NEGATED_OPERATORS
+    if attributes:
         for key, value_list in attributes.items():
             for idx, value in enumerate(value_list):
-                params["{}_{}_{}_attributes_regex".format(prepend, key, idx)] = ".*?({}).*?".format(
-                    ".*?".join(['{}="{}"'.format(key, value)])
-                )
-                or_conditions.append("match(elements_chain, %({}_{}_{}_attributes_regex)s)".format(prepend, key, idx))
-            if len(or_conditions) > 0:
-                conditions.append("(" + (" OR ".join(or_conditions)) + ")")
+                optional_flag = "(?i)" if operator.endswith("icontains") else ""
+                params[f"{prepend}_{key}_{idx}_attributes_regex"] = f'{optional_flag}({key}="{value}")'
+                conditions.append(f"match(elements_chain, %({prepend}_{key}_{idx}_attributes_regex)s)")
+            if len(conditions) > 1:
+                conditions.append(f"({' AND '.join(conditions)})")
+    if conditions:
+        final_conditions = f"{'NOT ' if is_negated else ''}({' OR '.join(conditions)})"
+    else:
+        final_conditions = "" if is_negated else "0 = 192"
```

Comment:
This does not match the previous behavior - this is a new bug!

Reply:
I don't think I see the bug, could you set out an example?

Reply:
The behavior changed when calling `filter_element` from `format_action_filter` on an autocapture event. Having no conditions is completely valid there - previously this function returned an empty string, now it returns `0 = 1`

Reply:
Right – made things smarter.

Reply:
+1 good catch re behavior in models/property here :sweat_smile: